plugins/permission-node/src/integration/createPermissionIntegrationRouter.ts

```diff
@@ -250,6 +250,108 @@ export type PermissionIntegrationRouterOptions<
   >;
 };
 
+class PermissionIntegrationMetadataStore {
+  readonly #rulesByTypeByName = new Map<
+    string,
+    Map<string, PermissionRule<unknown, unknown, string>>
+  >();
+  readonly #permissionsByName = new Map<string, Permission>();
+  readonly #resourcesByType = new Map<
+    string,
+    CreatePermissionIntegrationRouterResourceOptions<string, unknown>
+  >();
+  readonly #serializedRules = new Array<MetadataResponseSerializedRule>();
+
+  getSerializedMetadata(): MetadataResponse {
+    return {
+      permissions: Array.from(this.#permissionsByName.values()),
+      rules: this.#serializedRules,
+    };
+  }
+
+  hasResourceType(type: string): boolean {
+    return this.#resourcesByType.has(type);
+  }
+
+  async getResources(
+    resourceType: string,
+    refs: string[],
+  ): Promise<Record<string, unknown>> {
+    const resource = this.#resourcesByType.get(resourceType);
+    if (!resource?.getResources) {
+      throw new NotImplementedError(
+        `This plugin does not expose any permission rule or can't evaluate the conditions request for ${resourceType}`,
+      );
+    }
+
+    const uniqueRefs = Array.from(new Set(refs));
+    const resources = await resource.getResources(uniqueRefs);
+    return Object.fromEntries(
+      uniqueRefs.map((ref, index) => [ref, resources[index]]),
+    );
+  }
+
+  getRuleMapper(resourceType: string) {
+    return (name: string): PermissionRule<unknown, unknown, string> => {
+      const rule = this.#rulesByTypeByName.get(resourceType)?.get(name);
+      if (!rule) {
+        throw new Error(`Unexpected permission rule: ${name}`);
```

Comment:
have the resource type as well in this message perhaps?

Reply:
Copied from https://github.com/backstage/backstage/blob/eb9252186fb5a237eb389df716bc5094e54b9acc/plugins/permission-node/src/integration/util.ts#L81 so inclined to keep as is